src/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.Http2.cs

```diff
@@ -0,0 +1,243 @@
+            {
+                Task sendTask = client.GetAsync(server.CreateServer());
+
+                await server.AcceptConnectionAsync();
+                await server.SendConnectionPrefaceAsync();
+
+                // Receive the initial settings frame from the server.
+                Frame receivedFrame = await server.ReadFrameAsync(TimeSpan.FromSeconds(30));
+
+                // Receive the request header frame.
+                receivedFrame = await server.ReadFrameAsync(TimeSpan.FromSeconds(30));
+
+                // Send a malformed frame.
+                DataFrame invalidFrame = new DataFrame(new byte[0], FrameFlags.Padded, 10, 1);
+                await server.WriteFrameAsync(invalidFrame);
+
+                await Assert.ThrowsAsync<HttpRequestException>(async () => await sendTask);
+
+                // Receive a RST_STREAM frame.
+                receivedFrame = await server.ReadFrameAsync(TimeSpan.FromSeconds(30));
+                Assert.Equal(FrameType.RstStream, receivedFrame.Type);
+            }
+        }
+
+        [Fact]
+        [ActiveIssue(31514)]
+        public async Task ClosedStream_FrameReceived_ResetsStream()
+        {
+            HttpClientHandler handler = CreateHttpClientHandler();
+            handler.ServerCertificateCustomValidationCallback = TestHelper.AllowAllCertificates;
+
+            using (var server = new Http2LoopbackServer(new Http2Options()))
+            using (var client = new HttpClient(handler))
+            {
+                Task sendTask = client.GetAsync(server.CreateServer());
+
+                await server.AcceptConnectionAsync();
+                await server.SendConnectionPrefaceAsync();
+
+                // Receive the initial settings frame from the server.
+                Frame receivedFrame = await server.ReadFrameAsync(TimeSpan.FromSeconds(30));
+
+                // Receive the request header frame.
+                receivedFrame = await server.ReadFrameAsync(TimeSpan.FromSeconds(30));
+
+                // Send a reset stream frame so that stream 1 moves to a terminal state.
+                RstStreamFrame resetStream = new RstStreamFrame(FrameFlags.Padded, 0x1, 1);
+                await server.WriteFrameAsync(resetStream);
+
+                // Send a frame on the now-closed stream.
+                DataFrame invalidFrame = new DataFrame(new byte[10], FrameFlags.Padded, 10, 1);
+                await server.WriteFrameAsync(invalidFrame);
+
+                // Receive a RST_STREAM frame.
+                receivedFrame = await server.ReadFrameAsync(TimeSpan.FromSeconds(30));
+                Assert.Equal(FrameType.RstStream, receivedFrame.Type);
+
+                await Assert.ThrowsAsync<HttpRequestException>(async () => await sendTask);
+            }
+        }
+
+        public SocketsHttpHandler_HttpClientHandler_Http2_Test()
+        {
+            AppContext.SetSwitch("System.Net.Http.SocketsHttpHandler.Http2Support", true);
```

Comment:
I'm not sure that this approach for enabling HTTP/2 in the tests is safe. Since the setting is process wide, I'm worried about these tests running in parallel with (and breaking) the HTTP/1.1 tests.

Reply:
Have you run this yourself running all tests to see what happens?

```
cd src\System.Net.Http\tests\FunctionalTests
msbuild /t:rebuildandtest /p:Outerloop=true
```

This will turn on the AppContext switch for the test process and I expect other tests will fail because some HTTP/2 features aren't working yet (i.e. request body). So, when those tests run against the HTTP/2 server we have they will fail (i.e. PostScenarioTest.cs).

I don't think we are ready yet to turn on the AppContext switch.  And when we do, we need to do it in a way similar to how SocketsHttpHandlerTest wraps the abstract Http*Test classes.

Reply:
I have, though the tests haven't failed. I expect that it is timing dependent. I'll mark this as no merge until we figure out exactly how to do the testing. I'd still like to get feedback on the other elements of the change.

Reply:
Looks like your tests are failing.  And it's because the logic to turn on/off the AppContext is being done per-test which isn't good since all the tests in a class run in parallel.

Instead, you should switch to using a per-class initializer for this test class.
https://xunit.github.io/docs/shared-context

Reply:
Back before we made SocketsHttpHandler the default, we had some code in the tests for enabling this for testing purposes.  @stephentoub implemented this.  I don't remember the details, but I believe it was thread-safe.  I expect we can do something similar here.

If you dig through the SocketsHttpHandler test infra code history, you should be able to find it.  If you're having trouble, let me know and I can see if I can find it.

Reply:
It was based on using thread data.

https://github.com/dotnet/corefx/commit/852d40e723c849e8d0c486676d40332343e12ca5#diff-a8a18c63e3d6506e3f82b3f15ab9a47a

I'm hoping we avoid having to do this.  I'm considering putting back some internal property on the SocketsHttpHandler for now that we can set/unset via reflection.

But once we implement request body upload, then we can just set HTTP/2 enabled feature for SocketsHttpHandler everywhere in our tests.

Reply:
Using reflection seems fine to me.  I can't recall why we didn't do that for SocketsHttpHandler in the first place...

Reply:
@davidsh You're right, it must be the dispose. I'll try switching to the IClassFixture.

Using reflection sounds good to me though. I think it will be more reliable than setting the AppContextSwitch process wide.

Reply:
We can't document to customers to use reflection to turn off HTTP/2 support.  We either have a public API or AppContext/EnvVars.  We decided on the AppContext/EnvVars approach.

But for our own internal testing, we probably can put back the internal SocketsHttpHandler.MaxHttpVersion property.  And then use it in some fashion via reflection to help make our tests have more granular control over this.

I will look into this some more today/tomorrow and perhaps submit a PR to help here.